### PR TITLE
Feat(eos_designs): Add option to disable default 'redistribute connected' in VRF.

### LINKED ISSUE #4091: Add option to disable  default 'redistribute connected ' in VRF.
### Enhancement summary

Add option to disable  default 'redistribute connected ' in VRF.

### Which component of AVD is impacted

eos_designs, eos_cli_config_gen

### Use case example

MPLS topology . Customer does not want to see connected routes in RT.

### Describe the solution you would like

We need an option per VRF to disable 'redistribute connected'.

### Describe alternatives you have considered

_No response_

### Additional context

_No response_

### Contributing Guide

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by laxmikantchintakindi**
## Change Summary

Add option to disable default 'redistribute connected ' in VRF.

## Related Issue(s)

Fixes #4091
## Component(s) name

`arista.avd.eos_designs`

## Proposed changes
Add option to disable default 'redistribute connected ' in VRF. MPLS topology . Customer does not want to see connected routes in RT.

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [ ] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py [hunk 1]
```diff
@@ -304,7 +305,8 @@ def _update_router_bgp_vrf_mlag_neighbor_cfg(self: AvdStructuredConfigNetworkSer
         In-place update MLAG neighbor part of structured config for *one* VRF under router_bgp.vrfs
         """
         if not self._mlag_ibgp_peering_redistribute(vrf, tenant):
-            bgp_vrf["redistribute_routes"][0]["route_map"] = "RM-CONN-2-BGP-VRFS"
+            if len(bgp_vrf["redistribute_routes"]) > 0:
```
Comment (Shivani-gslab) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]:
```suggestion
            if bgp_vrf["redistribute_routes"]:
```

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_ospf.py [hunk 2]
```diff
@@ -76,10 +76,10 @@ def router_ospf(self: AvdStructuredConfigNetworkServices) -> dict | None:
                     if (route_map := get(vrf, "ospf.redistribute_bgp.route_map")) is not None:
                         process_redistribute["bgp"]["route_map"] = route_map
 
-                if get(vrf, "ospf.redistribute_connected.enabled", default=False) is True:
-                    process_redistribute["connected"] = {}
-                    if (route_map := get(vrf, "ospf.redistribute_connected.route_map")) is not None:
-                        process_redistribute["connected"]["route_map"] = route_map
+                    if get(vrf, "ospf.redistribute_connected.enabled", default=False) is True and get(vrf, "redistribute_connected", default=True) is True:
```
Comment (Shivani-gslab) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/router_ospf.py]: I think, the indentation is wrong here

Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/router_ospf.py]: Fixed.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py (hunk 1 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]: changed condition.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_ospf.py [hunk 3]
```diff
@@ -76,7 +76,7 @@ def router_ospf(self: AvdStructuredConfigNetworkServices) -> dict | None:
                     if (route_map := get(vrf, "ospf.redistribute_bgp.route_map")) is not None:
                         process_redistribute["bgp"]["route_map"] = route_map
 
-                if get(vrf, "ospf.redistribute_connected.enabled", default=False) is True:
+                if get(vrf, "ospf.redistribute_connected.enabled", default=False) is True and get(vrf, "redistribute_connected", default=True) is True:
```
Comment (MaheshGSLAB) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/router_ospf.py]: This become confusing as first condition have default as False and second condition have default as True for same config.
SO can change the condition as 
if get(vrf, "redistribute_connected", get(vrf, "ospf.redistribute_connected.enabled", default=False)) is True:

Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/router_ospf.py]: Fixed.

Comment (ClausHolbechArista) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/router_ospf.py]: I don't think we should blend these two knobs. Maybe the new knob should be under BGP and only control that.

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml [hunk 4]
```diff
@@ -897,6 +897,10 @@ $defs:
               redistribute_static:
                 type: bool
                 description: Non-selectively enabling or disabling redistribute static inside the VRF.
+              redistribute_connected:
+                type: bool
+                default: true
+                description: Enabled/disable redistribution of connected routes.
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]:
```suggestion
                description: Non-selectively enabling or disabling redistribution of connected routes to BGP.
```

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml [hunk 5]
```diff
@@ -897,6 +897,10 @@ $defs:
               redistribute_static:
                 type: bool
                 description: Non-selectively enabling or disabling redistribute static inside the VRF.
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]:
```suggestion
                description: Non-selectively enabling or disabling redistribute static to BGP inside the VRF.
```

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_ospf.py [hunk 6]
```diff
@@ -76,7 +76,7 @@ def router_ospf(self: AvdStructuredConfigNetworkServices) -> dict | None:
                     if (route_map := get(vrf, "ospf.redistribute_bgp.route_map")) is not None:
                         process_redistribute["bgp"]["route_map"] = route_map
 
-                if get(vrf, "ospf.redistribute_connected.enabled", default=False) is True:
+                if get(vrf, "redistribute_connected", get(vrf, "ospf.redistribute_connected.enabled", default=False)) is True:
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/router_ospf.py]: Suggesting to revert this change so we keep BGP and OSPF logic separate.
```suggestion
                if get(vrf, "ospf.redistribute_connected.enabled", default=False) is True:
```

**Review by ClausHolbechArista — CHANGES_REQUESTED** [review]

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml (hunk 4 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]: Fixed.

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml (hunk 5 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]: FIxed.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_ospf.py (hunk 6 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/router_ospf.py]: FIxed.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_ospf.py (hunk 3 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/router_ospf.py]: Fixed this under the comment https://github.com/aristanetworks/avd/pull/4220#discussion_r1692520763

**Review by ClausHolbechArista — APPROVED** [review]

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py [hunk 7]
```diff
@@ -168,11 +168,15 @@ def _router_bgp_vrfs(self: AvdStructuredConfigNetworkServices) -> list | None:
...
+                bgp_vrf_redistribute_connected = get(vrf, "redistribute_connected", default=True)
                 if vrf_name != "default":
                     bgp_vrf |= {
                         "router_id": self.shared_utils.router_id,
-                        "redistribute_routes": [{"source_protocol": "connected"}],
                     }
+                    if bgp_vrf_redistribute_connected is True:
+                        bgp_vrf |= {
+                            "redistribute_routes": [{"source_protocol": "connected"}],
+                        }
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]:
```suggestion
                    bgp_vrf["router_id"] = self.shared_utils.router_id
                    if bgp_vrf_redistribute_connected is True:
                        bgp_vrf["redistribute_routes"] = [{"source_protocol": "connected"}]
```

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py [hunk 8]
```diff
@@ -303,7 +307,7 @@ def _update_router_bgp_vrf_mlag_neighbor_cfg(self: AvdStructuredConfigNetworkSer
         """
         In-place update MLAG neighbor part of structured config for *one* VRF under router_bgp.vrfs
         """
-        if not self._mlag_ibgp_peering_redistribute(vrf, tenant):
+        if not self._mlag_ibgp_peering_redistribute(vrf, tenant) and get(vrf, "redistribute_connected", default=True):
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]: Would be best to have `self._mlag_ibgp_peering_redistribute` looking for `get(vrf, "redistribute_connected", default=True)` and setting the value based on this (move the and in the self._mlag...)

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml [hunk 9]
```diff
@@ -896,7 +896,11 @@ $defs:
                         type: str
               redistribute_static:
                 type: bool
-                description: Non-selectively enabling or disabling redistribute static inside the VRF.
+                description: Non-selectively enabling or disabling redistribute static to BGP inside the VRF.
+              redistribute_connected:
+                type: bool
+                default: true
+                description: Non-selectively enabling or disabling redistribution of connected routes to BGP.
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]:
```suggestion
                description: Non-selectively enabling or disabling redistribution of connected routes to BGP inside the VRF.
```

**Review by gmuloc — CHANGES_REQUESTED** [review]

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py (hunk 8 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]: Fixed.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py (hunk 7 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]: Fixed.

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml (hunk 9 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]: Added.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py [hunk 10]
```diff
@@ -303,8 +303,9 @@ def _update_router_bgp_vrf_mlag_neighbor_cfg(self: AvdStructuredConfigNetworkSer
         """
         In-place update MLAG neighbor part of structured config for *one* VRF under router_bgp.vrfs
         """
-        if not self._mlag_ibgp_peering_redistribute(vrf, tenant):
-            bgp_vrf["redistribute_routes"][0]["route_map"] = "RM-CONN-2-BGP-VRFS"
+        if get(vrf, "redistribute_connected", default=True):
+            if not self._mlag_ibgp_peering_redistribute(vrf, tenant):
+                bgp_vrf["redistribute_routes"][0]["route_map"] = "RM-CONN-2-BGP-VRFS"
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]: Guillaume meant to move the check for `get(vrf, "redistribute_connected", default=True)` into the method `def _mlag_ibgp_peering_redistribute(...)` So anything else using this method will honor the same.

**Review by ClausHolbechArista — CHANGES_REQUESTED** [review]

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py (hunk 10 above)
Comment (ClausHolbechArista) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]: Note that you must return True from that method in the case redistribute-connected is set to false since the check here has the NOT. Maybe we should reverse the logic and rename the method to _exclude_mlag_ibgp_peering_from_redistribute. Please analyze the logic and search for other uses of the method to see that you retain the current behavior.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py [hunk 11]
```diff
@@ -303,6 +303,7 @@ def _update_router_bgp_vrf_mlag_neighbor_cfg(self: AvdStructuredConfigNetworkSer
         """
         In-place update MLAG neighbor part of structured config for *one* VRF under router_bgp.vrfs
         """
+
```
Comment (laxmikantchintakindi) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]:
```suggestion
```

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py (hunk 10 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]: Fixed.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py [hunk 12]
```diff
@@ -162,15 +162,18 @@ def _mlag_ibgp_peering_vlan_vrf(self: AvdStructuredConfigNetworkServices, vrf: d
 
         return vlan_id
 
-    def _mlag_ibgp_peering_redistribute(self: AvdStructuredConfigNetworkServices, vrf: dict, tenant: dict) -> bool:
+    def _exclude_mlag_ibgp_peering_from_redistribute(self: AvdStructuredConfigNetworkServices, vrf: dict, tenant: dict) -> bool:
         """
         Returns True if MLAG IBGP Peering subnet should be redistributed for the given vrf/tenant.
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py]:
```suggestion
        Returns True if MLAG IBGP Peering subnet should be _excluded_ from redistribution for the given vrf/tenant.
```

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py (hunk 10 above)
Comment (ClausHolbechArista) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]: There is a bug in the new logic. If you try to set the new value to false on the VRF Tenant_A_WEB_ZONE you should see an error around the line mentioned above.

**Review by ClausHolbechArista — DISMISSED** [review]

File: python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py (hunk 12 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py]: Corrected the docstring.

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py (hunk 10 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]: _exclude_mlag_ibgp_peering_redistribute was returning True when `redistribute_connected` was false. Now changed it to return false. This is not giving the error for Tenant_A_WEB_ZONE when redistribute_connected is set to false.

Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]: Thanks.

Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]: Added a new vrf in network services to test this condition.

**Review by MaheshGSLAB — APPROVED** [review]

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml [hunk 13]
```diff
@@ -892,7 +892,11 @@ $defs:
                         type: str
               redistribute_static:
                 type: bool
-                description: Non-selectively enabling or disabling redistribute static inside the VRF.
+                description: Non-selectively enabling or disabling redistribute static to BGP inside the VRF.
+              redistribute_connected:
+                type: bool
+                default: true
+                description: Non-selectively enabling or disabling redistribution of connected routes to BGP inside the VRF.
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]:
```suggestion
                description: Enable or disable the redistribution of all connected routes to BGP in the VRF.
```

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml [hunk 14]
```diff
@@ -892,7 +892,11 @@ $defs:
                         type: str
               redistribute_static:
                 type: bool
-                description: Non-selectively enabling or disabling redistribute static inside the VRF.
+                description: Non-selectively enabling or disabling redistribute static to BGP inside the VRF.
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]:
```suggestion
                description: Enable or disable the redistribution of all static routes to BGP in the VRF.
```

Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]: Added this in latest commit.

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml (hunk 13 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]: Added this in latest commit.

**Review by emilarista — APPROVED** [review]
Tested, LGTM!

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py [hunk 15]
```diff
@@ -165,12 +165,12 @@ def _router_bgp_vrfs(self: AvdStructuredConfigNetworkServices) -> dict:
...
                     self._update_router_bgp_vrf_evpn_or_mpls_cfg(bgp_vrf, vrf, vrf_address_families)
 
+                bgp_vrf_redistribute_connected = get(vrf, "redistribute_connected", default=True)
                 if vrf_name != "default":
                     # Non-default VRF
-                    bgp_vrf |= {
-                        "router_id": self.shared_utils.router_id,
-                        "redistribute_routes": [{"source_protocol": "connected"}],
-                    }
+                    bgp_vrf |= {"router_id": self.shared_utils.router_id}
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]:
```suggestion
                    bgp_vrf["router_id] = self.shared_utils.router_id
```

File: python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py [hunk 16]
```diff
@@ -165,12 +165,12 @@ def _router_bgp_vrfs(self: AvdStructuredConfigNetworkServices) -> dict:
...
+                    if bgp_vrf_redistribute_connected is True:
+                        bgp_vrf |= {"redistribute_routes": [{"source_protocol": "connected"}]}
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/router_bgp.py]:
```suggestion
                        bgp_vrf["redistribute_routes"] = [{"source_protocol": "connected"}]
```

**Review by gmuloc — COMMENTED** [review]
small refactoring comments - otherwise looks good

**Review by gmuloc — APPROVED** [review]

**Review by gmuloc — APPROVED** [review]

**Review by bjmeuer — APPROVED** [review]
LGTM

**Comment by ClausHolbechArista** [comment]
@carlbuchmann I was blocking this PR since it needed more inspection like we talked about. It is not handling VRF default if configured under network services. We need to call that out in the docs. Like "only applicable to non-default VRFs".
